### PR TITLE
ColorListPicker: refactor to TypeScript

### PR DESCRIPTION
**Pull request by chad1008**
## What?
Refactor `ColorListPicker` component to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744

## Why?

The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

## Testing Instructions
- Ensure all tests still pass
- Confirm Storybook examples and editor examples work as expected. For now, it can be found/tested in components like the Duotone filter.

### Testing Instructions for Keyboard

TBD

### CODE REVIEW

**Comment by chad1008** [comment]
## HALP!

Spot where I'm currently stuck:

When typing the `onChange` methods for `ColorListPicker` and its subcomponent `ColorOption` I have a conflict I can't quite shake.
- `ColorListPicker`'s `value` prop is typed as `Array< string >`, and its `onChange` method expects an `Array< string >` arg.
- `ColorOption`'s `onChange` accepts a single `string` value. It copies the `value` array mentioned above and appends the new string to it. That new array is passed to `ColorListPicker`'s `onChange`. This is fine, it's still an array of strings.
- `ColorOption`'s onChange is forwarded on to `ColorPalette`. Unfortunately, `ColorPalette`'s `onChange` wants the string we're passing to be optional - meaning we have a conflict between `string` and a potential `undefined`.

Potential solution:
I'm wondering if the problem isn't with that optional argument in `ColorPalette`, and if it shouldn't, in fact, be a required `string | undefined` argument. [Further down the line in `ColorPalette`](https://github.com/WordPress/gutenberg/blob/0b1998cf997d866bc1eca182fedb1abb42e337fd/packages/components/src/color-palette/index.tsx#L231), `clearColor` is written to explicitly pass `undefined` to `onChange`, so we need to account for that flow. If we replace `undefined` with an empty string, and make the first `onChange` arg required, everything lines up nicely.

This also feels like the right move to me, because the first argument isn't actually optional - there's a second argument (`index`) to `onChange`,  and the order matters, so you really can't just leave `newColor` out. You'd need to pass _something_ there, to avoid colliding with `index`.

I began working on a PR to demo this change, but it had a bit of a domino effect that spills over to other components, so I put a pin in that for now. I'm happy to chase it further if this feels like the right road to travel!

File: packages/components/src/color-list-picker/types.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	onChange: ( newValue: Array< string > ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
+};
+
+export type ColorOptionProps = Pick<
+	ColorListPickerProps,
+	'colors' | 'disableCustomColors' | 'enableAlpha'
+> & {
+	label: ColorListPickerProps[ 'labels' ][ number ];
+	value: string; // TODO: can we extract this from ColorListPickerProps['value']? That prop being optional is a little tricky
+	onChange: ( newValue: string ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
```
Comment (mirka) [inline packages/components/src/color-list-picker/types.ts]: I'm not sure if I fully understand the potential solution you described, but I'm assuming it's something along these lines? If so, yes I think it makes sense!

```suggestion
	value?: Array< string | undefined >;
	disableCustomColors?: boolean;
	enableAlpha?: boolean;
	onChange: ( newValue: Array< string | undefined > ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
};

export type ColorOptionProps = Pick<
	ColorListPickerProps,
	'colors' | 'disableCustomColors' | 'enableAlpha'
> & {
	label: ColorListPickerProps[ 'labels' ][ number ];
	value: string | undefined; // TODO: can we extract this from ColorListPickerProps['value']? That prop being optional is a little tricky
	onChange: ( newValue: string | undefined ) => void; // TODO: resolve conflict with ColorPalette's onChange prop, which expects an optional value arg.
```

**Review by mirka — COMMENTED** [review]
Since this component is not exported publicly and is apparently only used by DuotonePicker, it might be simpler to just move this into the `duotone-picker/` folder and type it as part of DuotonePicker. We can always extract it back out when necessary.

Btw I don't mind moving forward with this PR as is (without moving it into DuotonePicker) as an intermediary step. I just wanted to say that it's probably not worth the effort to write new stories/tests/docs for this internal-only component.

File: packages/components/src/color-list-picker/types.ts (hunk 1 above)
Comment (chad1008) [reply packages/components/src/color-list-picker/types.ts]: Thanks @mirka! I'm getting back into this after being out for the holidays. I did consider the `| undefined` approach, but also (as I tend to do) dove down another rabbit hole at the same time.

The tl;dr of my previously mentioned solution was that the whole reason the `undefined` is needed is that another componenet (`ColorPalette`) explicitly passes an `undefined` value` when de-selecting a color. I think we can use an empty sting instead to both simplify the typing, and prevent a potential edge case bug that TypeScript might not currently see for us, involving an optional parameter that may not be safe to actually omit in all scenarios). More thorough description and a potential change in #46890 😄 

If that's not a good direction though, I'm perfectly happy with your proposed solution of adding `| undefined` as needed!

**Comment by chad1008** [comment]
> Btw I don't mind moving forward with this PR as is (without moving it into DuotonePicker) as an intermediary step. I just wanted to say that it's probably not worth the effort to write new stories/tests/docs for this internal-only component.

Thanks for that, I'll skip those TODOs and we can plan on rolling this into DuotonePicker in a follow up!

File: packages/components/src/color-list-picker/types.ts (hunk 1 above)
Comment (chad1008) [reply packages/components/src/color-list-picker/types.ts]: Based on the feedback from @jorgefilipecosta on #46890, I've scrapped that approached and gone with `string | undefined` here.

I've also added some JSDoc descriptions to types. Probably not mission-critical on an internal component like this one but I figured I may as well include them while I was updating things.

**Review by mirka — APPROVED** [review]
Looks good 🚀

**Pull request merged**